recipes/lora_finetune_distributed_td.py

```diff
@@ -0,0 +1,1150 @@
+                    lambda *args: noop_ctx.__exit__(), always_call=True
+                )
+
+        if self._is_rank_zero:
+            log.info(
+                f"Instantiating model and loading checkpoint took {time.perf_counter() - init_start:.2f} secs"
+            )
+            memory_stats = training.get_memory_stats(device=self._device)
+            training.log_memory_stats(memory_stats)
+
+        # synchronize before training begins
+        torch.distributed.barrier()
+
+        return model
+
+    def _setup_optimizer(
+        self, cfg_optimizer: DictConfig, opt_state_dict: Optional[Dict[str, Any]] = None
+    ) -> Optimizer:
+        optimizer = config.instantiate(cfg_optimizer, self._model.parameters())
+        if opt_state_dict:
+            training.load_from_full_optimizer_state_dict(
+                optimizer,
+                opt_state_dict,
+                self._device,
+            )
+
+        if self._is_rank_zero:
+            log.info("Optimizer is initialized.")
+        return optimizer
+
+    def _setup_lr_scheduler(
+        self,
+        cfg_lr_scheduler: DictConfig,
+        num_training_steps: int,
+        last_epoch: int,
+    ) -> Optimizer:
+        lr_scheduler = config.instantiate(
+            cfg_lr_scheduler,
+            self._optimizer,
+            num_training_steps=num_training_steps,
+            last_epoch=last_epoch,
+        )
+        if self._is_rank_zero:
+            log.info("Learning rate scheduler is initialized.")
+        return lr_scheduler
+
+    def _setup_one_dataset(
+        self,
+        cfg_dataset: DictConfig,
+        global_shuffle: bool,
+        global_parallel_method: str,
+        global_streaming: bool,
+        global_num_workers: int,
+    ) -> BaseNode:
+        streaming = cfg_dataset.pop("streaming", global_streaming)
+        parallel_method = cfg_dataset.pop("parallel_method", global_parallel_method)
+        shuffle = cfg_dataset.pop("shuffle", global_shuffle)
+        num_workers = cfg_dataset.pop("num_workers", global_num_workers)
+
+        return config.instantiate(
+            cfg_dataset,
+            self._tokenizer,
+            streaming=streaming,
+            shuffle=shuffle,
+            parallel_method=parallel_method,
+            num_workers=num_workers,
+        )
+
+        return ds
+
+    def _setup_data_td(
+        self,
+        cfg_dataloader: DictConfig,
+        cfg_multi_datasets: DictConfig,
+        batch_size: int,
+    ) -> Loader:
+        """
+        All torchdata related setup happens here. Currently this recipe supports
+        both Map and Streaming datasets (from HuggingFace datasets), and mixing multiple
+        datasets (can be mix of Map and Streaming).
+        """
+        world_size, rank = training.get_world_size_and_rank()
+
+        # Get global settings
+        shuffle = cfg_dataloader.shuffle
+        parallel_method = cfg_dataloader.get("parallel_method", "thread")
```

Comment:
The trouble is that there's no real distinction between datasets and dataloader with torchdata.nodes, they're all just Iterators that can be composed together. Currently it's exposed but we can introduce restrictions if you think it would be simpler